pkg/network/common/egressip.go

```diff
@@ -290,6 +300,153 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkv1.HostSubnet) {
 	eit.syncEgressIPs()
 }
 
+func nodeIsReady(node *v1.Node) bool {
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == v1.NodeReady {
+			if cond.Status == v1.ConditionFalse || cond.Status == v1.ConditionUnknown {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// SetMonitorNodes updates EgressIPTracker with the node data provided and
+// starts a go-routine (if one is not already running) which polls monitorNodes
+// to check if they are ready and reachable
+func (eit *EgressIPTracker) SetMonitorNodes(monitorNodes map[string]*NodeEgress) {
+	eit.Lock()
+	defer eit.Unlock()
+	eit.monitorNodes = monitorNodes
```

Comment:
This still has the problem I was talking about before with losing information. Every time a `NetNamespace` or `HostSubnet` gets updated and the master calls `SetMonitorNodes`, it will cause the tracker to forget the `sdnIP`, `requestedIPs`, `requestedCIDRs`, `parsedCIDRs`, `activeEgressIPs`, `offline`, and `retries` of all the nodes it had been tracking before.

You have to do a more complicated sort of "sync" operation here, adding elements to and removing elements from `eit.monitorNodes` so that at the end they have the same keys, but whenever there's overlap, it needs to keep the original copy from `eit.monitorNodes`, not replace it with the one from `monitorNodes`

Reply:
> This still has the problem I was talking about before with losing information.

No, I don't think what you're describing is correct. In the most recent, and current, version of this patch I added: https://github.com/openshift/sdn/blob/0551a495b59c87b6d5da5144a76e6ba2da4a717f/pkg/network/master/egressip.go#L90-L92 which means that `maybeDoUpdateEgressCIDRs` only does `SetMonitorNodes` for nodes not currently monitored. Since the access here: https://github.com/openshift/sdn/blob/0551a495b59c87b6d5da5144a76e6ba2da4a717f/pkg/network/master/egressip.go#L82  is thread-safe, there shouldn't be any problem w.r.t overriding data that's currently existing in `monitorNodes` 

I need will test what I am saying right now and confirm that this works.

Reply:
Actually, I think I see what you mean now. Indeed if any item in `monitorNodes` happens to be modified in between `GetMonitorNodes` and `SetMonitorNodes` in `maybeDoUpdateEgressCIDRs`, we will override the modified field with the old one. If we however instead do 

```go
if _, exists := monitorNodes[hs.HostIP]; !exists {  	
    eim.tracker.AddToMonitorNodes(&amp;common.NodeEgress{NodeIP: hs.HostIP, NodeName: nodeName})
} 
```
and have `AddToMonitorNodes` be thread-safe while removing `SetMonitorNodes`, we should be fine, right?

Reply:
Doing all of this I realized that `maybeDoUpdateEgressCIDRs` was inherently wrong. There was never any situation in which we would stop monitoring nodes (even if nodes are not marked for egress IP anymore). I've fixed this now, by implementing something that resembles the current implementation.